source/LiteralModel.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <mariana-trench/EventLogger.h>
+#include <mariana-trench/LiteralModel.h>
+#include <mariana-trench/Positions.h>
+
+namespace marianatrench {
+
+namespace {
+
+class LiteralModelConsistencyError {
+ public:
+  static void raise(const std::string& what) {
+    ERROR(1, "Literal Model Consistency Error: {}", what);
+    EventLogger::log_event("regex_model_consistency_error", what);
+  }
+};
+
+}
+
+static void check_taint_config_consistency(const std::string& pattern, const TaintConfig& frame) {
+  if (frame.kind() == nullptr) {
+    LiteralModelConsistencyError::raise(fmt::format(
+        "Model for regex `{}` must have a source kind.", pattern));
+  }
+}
+
+static Taint create_taint(const std::string& pattern, const std::vector<TaintConfig>& sources) {
+  Taint all_sources;
+  for (const auto& source : sources) {
+    mt_assert(source.is_leaf());
+    check_taint_config_consistency(pattern, source);
+    all_sources.join_with(Taint{std::move(source)});
+  }
+  return all_sources;
+}
+
+LiteralModel::LiteralModel() { }
+
+LiteralModel::LiteralModel(
+    std::string&& pattern,
+    const std::vector<TaintConfig>& sources)
+    : pattern_(std::move(pattern)), regex_(pattern_), sources_(create_taint(pattern_, sources)) {
+}
+
+const std::string& LiteralModel::pattern() const {
+  return pattern_;
+}
+
+const Taint& LiteralModel::sources() const {
+  return sources_;
+}
+
+bool LiteralModel::matches(const std::string_view literal) const {
+  return regex_match(std::cbegin(literal), std::cend(literal), regex_);
+}
+
+void LiteralModel::join_with(const LiteralModel& other) {
+  if (this != &other) {
+    sources_.join_with(other.sources_);
+  }
+}
+
+bool LiteralModel::empty() const {
+  return sources_.is_bottom();
+}
+
+static constexpr const char* PATTERN = "pattern";
+
+static constexpr const char* SOURCES = "sources";
+
+LiteralModel LiteralModel::from_json(
+    const Json::Value& value,
+    Context& context) {
+  JsonValidation::validate_object(value);
+  JsonValidation::check_unexpected_members(
+      value, {PATTERN, SOURCES, "description"});
+
+  std::vector<TaintConfig> sources;
+  for (auto source : JsonValidation::null_or_array(value, /* field */ SOURCES)) {
+    sources.push_back(TaintConfig::from_json(source, context));
+  }
+  return LiteralModel(JsonValidation::string(value, PATTERN), sources);
+}
+
+Json::Value LiteralModel::to_json(const ExportOriginsMode export_origins_mode) const {
+  auto value = Json::Value(Json::objectValue);
+
+  value[PATTERN] = pattern_;
+
+  if (!sources_.is_bottom()) {
+    auto sources_value = Json::Value(Json::arrayValue);
+    for (const auto& source : sources_.frames_iterator()) {
+      mt_assert(!source.is_bottom());
```

Comment:
similar to the above, I think this could be reported as a user error?

Reply:
This case is in `LiteralModel::to_json`, so there is no JSON user input being parsed here. This can only occur in the case of an application logic error, where we constructed a `LiteralModel` which cannot be serialised. After a quick browse I couldn't find any `::to_json` method throwing a `JsonValidationError`, I think they all use `mt_assert` in such cases.

However, we could e.g. throw an `std::invalid_argument` if preferred?

Reply:
ah yeah, my bad, I think `mt_assert` is good here, thanks!